### PR TITLE
Add refresh token 

### PR DESCRIPTION
**Pull request by cherylli**
# Description

1. Refresh token functionality is added. 
- refresh route
- refresh token in user table

2. Global guard to protect all the routes
- can be bypassed by `@Public` decorator

Refresh token will have 7 day expiry, while access token will expire in 15mins (I think access tokens can probably have a longer expiry)

Please add 2 new environment Variables
AT_SECRET
RT_SECRET

JWT_SECRET can be removed

## Issue link

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Tested locally

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by cherylli** [comment]
[Refresh token - hashing](https://trello.com/c/jFe6WiID/96-refresh-token-hashing)

**Comment by cherylli** [comment]
[Refresh token - Reuse detection](https://trello.com/c/2eMbgT8y/106-refresh-token-reuse-detection)

**Review by jenny-alexander — APPROVED** [review]
LGTM. I did various tests like:
- Access token and refresh token were granted after successful login.
- Access token was removed from cookies after 15 minutes.
- Refresh token was successfully granted after access token expires.
- I tampered with refresh token and when 'refresh' endpoint was invoked, it failed (as expected).
- Access token and refresh token were removed from cookies after successful logout & refreshToken set to null in database.

File: src/auth/auth.service.ts [hunk 1]
```diff
@@ -36,6 +39,39 @@ export class AuthService {
         return this.jwtService.sign(payload, { expiresIn: "1h" });
     };
 
+    // access token and refresh token
+    private generateAtRtTokens = async (payload: object) => {
+        const [at, rt] = await Promise.all([
+            this.jwtService.signAsync(payload, {
+                secret: process.env.AT_SECRET,
+                expiresIn: 60 * 60,
```
Comment (cherylli) [inline src/auth/auth.service.ts:47]: I forgot to change this to use the AT_MAX_AGE. do we want 15mins or 1h?

**Comment by cherylli** [comment]
> LGTM. I did various tests like:
> 
> * Access token and refresh token were granted after successful login.
> * Access token was removed from cookies after 15 minutes.
> * Refresh token was successfully granted after access token expires.
> * I tampered with refresh token and when 'refresh' endpoint was invoked, it failed (as expected).
> * Access token and refresh token were removed from cookies after successful logout & refreshToken set to null in database.

https://github.com/chingu-x/chingu-dashboard-be/pull/78/files#r1460185576

**Review by jenny-alexander — APPROVED** [review]

**Review by curtwl — APPROVED** [review]
LGTM

**Pull request merged**